기록 저장 화면에서 등록 버튼 클릭 후, CloudStorage 이미지 업로드 -> 파이어베이스 DB 업데이트를 백그라운드에서 순차 처리 및 프로그레스바 알림 구현

jungkeunking opened the pull request:
기록 저장 화면 등록 버튼 클릭시:
- [x] (주한님) WorkManager 활용해, 사용자가 추가한 사진들을 백그라운드에서 CloudStorage에 업로드
- [x] 추가적으로, 백그라운드에서 Firebase DB 업데이트 (기존 SaveLogViewModel 및 Fragment 내의 함수들 이전시켜 작업 수행)
- [x] CustomWorkerFactory 클래스를 구현해 Worker 인스턴스 직접 생성 (HiltWorker의 기본 팩토리로는 context, workerparams 외에 의존성을 추가적으로 주입하면 오류 발생 -> 디버깅 실패해서 직접 팩토리 구현)
- [x] 위의 두 백그라운드 작업들을 체이닝시켜 연쇄 호출
- [x] CloudStorage 여러 이미지들 업로드에 대해 통합된 프로그레스바 알림 구현
- [x] 등록 버튼 클릭시 알림 권한 요청 처리한 다음, 화면 전환하도록 구현 

2/17 추가 수정 사항 (Firebase Realtime DB images 테이블에 CloudStorage URL이 아닌 URI를 저장하는 오류 수정)
- [x] 기존 ImageModel 프로퍼티명 url -> uri로 변경 및 패키지 savelog -> model로 변경
- [x] 업로드된 이미지의 uri 대신 url을 Firebase Realtime DB에 저장하도록 구현 수정 및 해당 값을 ImageUploadWorker에서 DatabaseUpdateWorker로 전달하도록 구현 수정 

추가 수정 사항 반드시 확인해주시구요
제가 변수명 관해서 다들 최소 한번씩은 피드백 드린 거 같은데, 변수명 제대로 짓는 거 엄청 중요하니까 다들 더 신경 좀 써주세요
기존 구현에 url과 uri가 잘못 섞여 있어서 디버깅하는 데 많이 애먹었습니다

jungkeunking committed:
build: 알림 권한 추가

jungkeunking committed:
feat: 홈 액티비티에 알림 채널 추가 및 기록저장 화면에 저장 버튼 클릭시 알림 구현

juhan2103 committed:
build: workmanager dependency 추가

juhan2103 committed:
feat: 사진 업로드 백그라운드 임시 구현

jungkeunking committed:
feat: 기록 저장 화면에서 저장 버튼 클릭시, 알림 권한 요청 및 백그라운드에서 알림 전송 구현

jungkeunking committed:
feat: 기록 저장 화면 백그라운드 이미지 업로드 일괄 처리 및 프로그레스바 알림 구현

jungkeunking committed:
build: 힐트 워크매니저 라이브러리 추가

jungkeunking committed:
build: 매니페스트에 Worker 기본 팩토리 비활성화 추가

jungkeunking committed:
refactor: ImageUploadWorker 클래스를 힐트워커로 변경

jungkeunking committed:
feat: DatabaseUpdateWorker 클래스 추가 (기존 뷰모델 로직 처리)

TODO: 리팩토링

jungkeunking committed:
feat: CustomWorkerFactory 클래스 추가 및 TreasureHuntApplication 클래스에 WorkManager Provider 구현

jungkeunking committed:
feat: SaveLog 프래그먼트에서 CloudStorage 업로드 워크에 체이닝시켜 DB 업데이트 워크 추가

jungkeunking commented on app/src/main/java/com/treasurehunt/ui/savelog/CoroutineUploadWorker.kt:
```diff
@@ -0,0 +1,61 @@
+import com.treasurehunt.util.NOTIFICATION_ID
+import com.treasurehunt.util.NOTIFICATION_ID_STRING
+
+class CoroutineUploadWorker(
```
이 클래스는 더 이상 사용되지 않아서 안 봐주셔도 됩니다!
여기 예전에 알림 권한 처리 같은 게 포함돼 있는데 그거 아직 이전이 덜 돼서 놔뒀습니다

jsulim commented:
고생하셨습니다!

jungkeunking committed:
remove: CoroutineUploadWorker 삭제 (알림 작업 이전)

jungkeunking committed:
feat & refactor: 기록 저장 화면에서 등록 버튼 클릭시 알림 권한 요청 처리 후 화면 전환되도록 수정, Work Data 키값 상수로 관리, 함수 순서 정리 및 제거, 스트링 리소스 추출 등

jungkeunking committed:
refactor: SaveLogViewModel 함수 정리, getImage -> getImagePick 함수명 수정, setButtonState 함수 리턴값 제거

donghyun81 commented on app/src/main/java/com/treasurehunt/ui/savelog/ImageUploadWorker.kt:
```diff
@@ -16,23 +17,31 @@ import com.treasurehunt.BuildConfig
 
-class ImageUploadWorker(private val context: Context, params: WorkerParameters) :
-    CoroutineWorker(context, params) {
+@HiltWorker
```
힐트 적용해주신거 좋았습니다!

jungkeunking committed:
build: 매니페스트에서 WorkManager Content Provider만 비활성화하도록 수정

donghyun81 reviewed (commented):
구현 고생 많으셨습니다!

juhan2103 reviewed (commented):
@jungkeunking 엄청난 구현이네요! 진짜 수고 많으셨습니다!

juhan2103 commented:
![image](https://github.com/Sesac-team-03/TreasureHunt/assets/83435602/9dad1f02-6b82-44fb-b23e-a9f10d1cf5a5)
등록 버튼을 눌렀을 때 알림 권한을 물음과 동시에 강제 종료가 됩니다!
위에 사진은 강제 종료 됐을 때의 로그캣의 오류 메세지 입니다! 수정 부탁드립니다!

jungkeunking committed:
fix: 기본 notification builder 삭제 (setIcon 없다는 오류 해결)

jungkeunking committed:
fix: 매니페스트에서 Content Provider 비활성화 코드 수정 및 totalByteCount 연산 수정

1) WorkerInitializer만 비활성화하려니까 안 돼서 일단 전체 비활성화로 변경
2) 낮은 API 레벨에서 totalByteCount가 실제 bytesTransferred의 절반인 버그가 추후 fix된 것으로 보여 높은 API에 맞춰서 변경

jungkeunking commented:
> ![image](https://private-user-images.githubusercontent.com/83435602/304631829-9dad1f02-6b82-44fb-b23e-a9f10d1cf5a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc4OTU5MzEsIm5iZiI6MTcwNzg5NTYzMSwicGF0aCI6Ii84MzQzNTYwMi8zMDQ2MzE4MjktOWRhZDFmMDItNmI4Mi00NGZiLWIyM2UtYTlmMTBkMWNmNWE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE0VDA3MjcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwNjU2OWRlNWNmNWVlZDI5YzE4Y2QwZjZiNDIxZDczYmJlNzFkY2VjMTNhOTQ5ZjE1MGE1ZTBhYmI4MWZmMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YBmS4NK12AcJ2EStci_sFBMXiY0xx5GnWsYd8cK_VKI) 등록 버튼을 눌렀을 때 알림 권한을 물음과 동시에 강제 종료가 됩니다! 위에 사진은 강제 종료 됐을 때의 로그캣의 오류 메세지 입니다! 수정 부탁드립니다!

감사합니다
로그아이콘 오류랑 DB 업뎃 안되는 오류 해결해서 추가 커밋 푸시했습니다!

jungkeunking committed:
fix: strings.xml 빠진 tag 추가

jungkeunking committed:
fix: LogModel 인스턴스 생성시 프로퍼티 순서 변경 반영

jungkeunking committed:
refactor: ImageModel 프로퍼티명 url -> uri로 변경 및 패키지 savelog -> model로 변경

jungkeunking committed:
fix: 업로드된 이미지의 uri 대신 url을 Firebase Realtime DB에 저장하도록 구현 수정 및 해당 값을 ImageUploadWorker에서 DatabaseUpdateWorker로 전달하도록 구현 수정
